app/controllers/application_controller.rb

```diff
@@ -102,35 +52,42 @@ def process_index(base, serializer_class, opts: {}, default_sort: nil, allowed_s
     scope = policy_scope(base)
     scope = order(scope, default_sort, allowed_sort_keys) if default_sort
     scope = paginate(scope)
-    meta = index_meta(scope)
-    render :json => serializer_class.new(scope, { :meta => meta }.deep_merge(opts))
+    total_records = scope.limit(nil).offset(nil).count
+    meta = index_meta(scope, total_records)
+    links = index_links(scope, total_records)
+    render :json => serializer_class.new(scope, { meta: meta, links: links }.deep_merge(opts))
   end
 
-  def index_meta(scope)
+  def index_meta(scope, total_records)
     meta = {
-      :total => scope.count,
-      :per_page => scope.per_page,
-      :page => scope.current_page
+      :total => total_records
     }
-    meta[:limit] = scope.limit_value if scope.respond_to?(:limit_value)
-    meta[:offset] = scope.offset_value if scope.respond_to?(:offset_value)
+    meta[:limit] = scope.limit_value if scope.respond_to?(:limit_value) && scope.limit_value
+    meta[:offset] = scope.offset_value if scope.respond_to?(:offset_value) && scope.offset_value
     meta
   end
 
-  def render_unprocessable_entity(errors)
-    render :json => { :errors => errors }, :status => :unprocessable_entity
-  end
+  def index_links(_scope, total_records)
+    links = {
+      first: same_request_with_parameters(first_records_parameters),
+      last: same_request_with_parameters(last_records_parameters(total_records))
+    }
 
-  def render_not_found(what, id)
-    message = "Could not find #{what}"
-    message += "with 'id'=#{id}" if id
-    render :json => { :errors => message }, :status => :not_found
+    offset = params[:offset]
+    if offset.to_i + limit_param < total_records
+      links[:next] = same_request_with_parameters(next_records_parameters)
+    end
+    links[:previous] = same_request_with_parameters(previous_records_parameters) if offset
```

Comment:
Requesting a page with `limit=10` and `offset=0` leads to `previous` being `http://localhost:3000/api/hooks/endpoints?limit=0&offset=0` which isn't particularly useful